presto-main/src/main/java/io/prestosql/server/security/UserExtraction.java

```diff
@@ -0,0 +1,134 @@
+            throws IOException
+    {
+        if (userExtractionPattern.isPresent()) {
+            checkArgument(!userExtractionFile.isPresent(), "user extraction pattern and file can not both be set");
+            return new UserExtraction(ImmutableList.of(new Rule(userExtractionPattern.get())));
+        }
+        if (userExtractionFile.isPresent()) {
+            List<Rule> rules = RULES_JSON_CODEC.fromJson(Files.readAllBytes(userExtractionFile.get().toPath())).getRules();
+            return new UserExtraction(rules);
+        }
+        return new UserExtraction(ImmutableList.of(new Rule("(.*)")));
+    }
+
+    @VisibleForTesting
+    UserExtraction(List<Rule> rules)
+    {
+        this.rules = ImmutableList.copyOf(requireNonNull(rules, "rules is null"));
+    }
+
+    public String extractUser(String principal)
+    {
+        Optional<String> user = rules.stream()
+                .map(rule -> rule.extractUser(principal))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .findFirst();
+
+        if (!user.isPresent()) {
+            throw new RuntimeException("No extraction patterns match the principal");
+        }
+        return user.get();
+    }
+
+    public static final class UserExtractionRules
+    {
+        private final List<Rule> rules;
+
+        @JsonCreator
+        public UserExtractionRules(
+                @JsonProperty("rules") List<Rule> rules)
+        {
+            this.rules = rules;
+        }
+
+        public List<Rule> getRules()
+        {
+            return rules;
+        }
+    }
+
+    public static final class Rule
+    {
+        private final Pattern pattern;
+        private final String user;
+        private final boolean allow;
+
+        public Rule(String pattern)
+        {
+            this(pattern, "$1", true);
+        }
+
+        @JsonCreator
+        public Rule(
+                @JsonProperty("pattern") String pattern,
+                @JsonProperty("user") Optional<String> user,
+                @JsonProperty("allow") Optional<Boolean> allow)
+        {
+            this(pattern, requireNonNull(user, "user is null").orElse("$1"), requireNonNull(allow, "allow is null").orElse(TRUE));
+        }
+
+        public Rule(String pattern, String user, boolean allow)
+        {
+            this.pattern = Pattern.compile(requireNonNull(pattern, "pattern is null"));
+            this.user = requireNonNull(user, "user is null");
+            this.allow = requireNonNull(allow, "allow is null");
+        }
+
+        public Optional<String> extractUser(String principal)
+        {
+            Matcher matcher = pattern.matcher(principal);
+            if (!matcher.matches()) {
+                return Optional.empty();
+            }
+            if (!allow) {
+                throw new RuntimeException("Principal is not allowed");
```

Comment:
Looks like authZ. Maybe it shouldn't be in this layer. Let's discuss.

Reply:
Discussed offline